src/lib/AST/AnyBlock.hpp

```diff
@@ -14,13 +14,195 @@
 #define MRDOX_LIB_AST_ANYBLOCK_HPP
 
 #include "BitcodeReader.hpp"
-#include "DecodeRecord.hpp"
-#include "lib/Support/Debug.hpp"
-#include "lib/Support/Error.hpp"
 
 namespace clang {
 namespace mrdox {
 
+// bool
+inline
+Error
+decodeRecord(
+    Record const& R,
+    bool& Field,
+    llvm::StringRef Blob)
+{
+    Field = R[0] != 0;
+    return Error::success();
+}
+
+// 32 bit integral types
+template<class IntTy>
+requires std::integral<IntTy>
+Error
+decodeRecord(
+    Record const& R,
+    IntTy& v,
+    llvm::StringRef Blob)
+{
+    v = 0;
+    if (R[0] > (std::numeric_limits<IntTy>::max)())
+        return formatError("integer overflow");
+    v = static_cast<IntTy>(R[0]);
+    return Error::success();
+}
+
+// integral types wider than 32 bits
+template<class IntTy>
+requires (std::integral<IntTy> &&
+    sizeof(IntTy) > 4)
+Error
+decodeRecord(
+    Record const& R,
+    IntTy& v,
+    llvm::StringRef Blob)
+{
+    v = static_cast<IntTy>(
+        static_cast<std::uint64_t>(R[0]) |
+        (static_cast<std::uint64_t>(R[1]) << 32));
+    return Error::success();
+}
+
+// enumerations
+template<class Enum>
+requires std::is_enum_v<Enum>
+Error
+decodeRecord(
+    Record const& R,
+    Enum& value,
+    llvm::StringRef blob)
+{
+    std::underlying_type_t<Enum> temp;
+    if(auto err = decodeRecord(R, temp, blob))
+        return err;
+    value = static_cast<Enum>(temp);
+    return Error::success();
+}
+
+// container of char
+template<class Field>
+requires std::is_same_v<
+    typename Field::value_type, char>
+Error
+decodeRecord(
+    const Record& R,
+    Field& f,
+    llvm::StringRef blob)
+        requires requires
+        {
+            f.assign(blob.begin(), blob.end());
+        }
+{
+    f.assign(blob.begin(), blob.end());
+    return Error::success();
+}
+
+// range<SymbolID>
+inline
+Error
+decodeRecord(
+    Record const& R,
+    std::vector<SymbolID>& f,
+    llvm::StringRef blob)
+{
+    auto src = R.begin();
+    auto n = *src++;
+    f.resize(n);
+    auto* dest = &f[0];
+    while(n--)
+    {
+        *dest++ = SymbolID(src);
+        src += BitCodeConstants::USRHashSize;
+    }
+    return Error::success();
+}
+
+inline
+Error
+decodeRecord(
+    const Record& R,
+    SymbolID& Field,
+    llvm::StringRef Blob)
+{
+    if (R[0] != BitCodeConstants::USRHashSize)
+        return formatError("USR digest size={}", R[0]);
+
+    Field = SymbolID(&R[1]);
+    return Error::success();
+}
+
+inline
+Error
+decodeRecord(
+    Record const& R,
+    OptionalLocation& Field,
+    llvm::StringRef Blob)
+{
+    if (R[0] > INT_MAX)
+        return formatError("integer value {} too large", R[0]);
+    Field.emplace((int)R[0], Blob, (bool)R[1]);
+    return Error::success();
```

Comment:
We don't have to worry about that in this PR, but the idiomatic way would be to return `Expected<void, Error>` and `return {}` on success instead of a somewhat contradictory and counterintuitive `Error::success()`.